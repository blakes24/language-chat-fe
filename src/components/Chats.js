import { makeStyles } from "@material-ui/core/styles";
import Typography from "@material-ui/core/Typography";
import { useDispatch, useSelector } from "react-redux";
import { fetchRooms, addRoom, setCurrentRoom } from "../store/roomSlice";
import { useParams } from "react-router-dom";
import ChatApi from "../helpers/api";
import { useEffect, useState } from "react";
import { Container, Divider, Paper } from "@material-ui/core";
import UserContext from "../helpers/UserContext";
import { useContext } from "react";
import RoomList from "./RoomList";

const useStyles = makeStyles((theme) => ({
  root: {
    padding: 0,
    margin: 0,
    paddingTop: ".1rem",
    height: "calc(100% - 56px)",
  },
  list: {
    width: 69,
    paddingBottom: 0,
    overflowX: "hidden",
    overflowY: "scroll",
    borderRight: "1px solid lightGray",
    height: "100%",
    maxHeight: "calc(100vh - 56px)",
    flexFlow: "column",
    display: "flex",
    [theme.breakpoints.up("md")]: {
      width: "250px",
      maxHeight: "calc(100vh - 64px)",
    },
  },
  chat: {
    height: "100%",
    display: "flex",
    width: "100%",
  },
  title: {
    padding: ".4rem 0",
  },
}));

function Chats() {
  const classes = useStyles();
  const { user } = useContext(UserContext);
  const { roomId } = useParams();
  const dispatch = useDispatch();
  const currentRoom = useSelector((state) => state.rooms.current);
  const rooms = useSelector((state) => state.rooms.items);
  const [error, setError] = useState(null);

  useEffect(() => {
    async function getData() {
      let ids;
      let partnerId;
      try {
        dispatch(fetchRooms({ userId: user.id }));
        if (roomId) {
          ids = roomId.split("-");
          partnerId = +ids[0] === user.id ? +ids[1] : +ids[0];
          let res = await ChatApi.getRooms({ userId: user.id, partnerId });
          dispatch(setCurrentRoom(res.rooms[0]));
        }
      } catch (err) {
        try {
          await ChatApi.createRoom({ user1: user.id, user2: partnerId });
          let res = await ChatApi.getRooms({ userId: user.id, partnerId });
          dispatch(addRoom(res.rooms[0]));
          dispatch(setCurrentRoom(res.rooms[0]));
        } catch {
          setError(err);
        }
      }
    }
    getData();
  }, [user, roomId, dispatch]);

  return (
    <Container className={classes.root}>
      <div className={classes.chat}>
        <Paper square className={classes.list}>
          {rooms.length > 0 && <RoomList rooms={rooms} />}
        </Paper>
        <Container className={classes.root}>
          <Typography
            className={classes.title}
            component={"h1"}
            variant="h4"
            align="center"
          >
            {currentRoom ? currentRoom.partner.name : "Chat"}
          </Typography>{" "}
          <Divider />
          {error && error.map((err) => <p>err</p>)}
        </Container>
      </div>
    </Container>
  );
}

export default Chats;
