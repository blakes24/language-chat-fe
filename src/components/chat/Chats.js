import { makeStyles } from "@material-ui/core/styles";
import { useDispatch, useSelector } from "react-redux";
import { fetchRooms, addRoom, setCurrentRoom } from "../../store/roomSlice";
import { useParams } from "react-router-dom";
import ChatApi from "../../helpers/api";
import { useEffect, useState } from "react";
import { Container, Divider, Typography } from "@material-ui/core";
import RoomList from "./RoomList";
import ChatRoom from "./ChatRoom";
import ChatAvatar from "./ChatAvatar";

const useStyles = makeStyles((theme) => ({
  root: {
    padding: 0,
    margin: 0,
    height: "calc(100% - 56px)",
    [theme.breakpoints.up("sm")]: {
      height: "calc(100% - 64px)",
    },
  },
  list: {
    width: 69,
    paddingBottom: 0,
    marginBottom: 0,
    overflowX: "hidden",
    overflowY: "scroll",
    borderRight: "1px solid lightGray",
    height: "100%",
    maxHeight: "calc(100vh - 56px)",
    flexFlow: "column",
    display: "flex",
    [theme.breakpoints.up("sm")]: {
      width: "250px",
      maxHeight: "calc(100vh - 64px)",
    },
  },
  chat: {
    height: "100%",
    display: "flex",
    width: "100%",
  },
  title: {
    padding: ".5rem 0",
    marginLeft: ".5rem",
  },
  chatContainer: {
    padding: 0,
    margin: 0,
    width: "100%",
  },
  partner: {
    marginTop: ".5rem",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    width: "100%",
    height: "4rem",
  },
}));

function Chats() {
  const classes = useStyles();
  const { roomId } = useParams();
  const dispatch = useDispatch();
  const user = useSelector((state) => state.users.current);
  const currentRoom = useSelector((state) => state.rooms.current);
  const rooms = useSelector((state) => state.rooms.items);
  const [error, setError] = useState(null);

  useEffect(() => {
    async function getData() {
      let ids;
      let partnerId;
      try {
        dispatch(fetchRooms({ userId: user.id }));
        if (roomId) {
          ids = roomId.split("-");
          partnerId = +ids[0] === user.id ? +ids[1] : +ids[0];
          let res = await ChatApi.getRooms({ userId: user.id, partnerId });
          dispatch(setCurrentRoom(res.rooms[0]));
        }
      } catch (err) {
        try {
          await ChatApi.createRoom({ user1: user.id, user2: partnerId });
          let res = await ChatApi.getRooms({ userId: user.id, partnerId });
          dispatch(addRoom(res.rooms[0]));
          dispatch(setCurrentRoom(res.rooms[0]));
        } catch {
          setError(err);
        }
      }
    }
    getData();
  }, [user, roomId, dispatch]);

  return (
    <Container maxWidth="xl" className={classes.root}>
      <div className={classes.chat}>
        <div className={classes.list}>
          {rooms.length > 0 && <RoomList rooms={rooms} />}
        </div>
        <Container className={classes.chatContainer}>
          {currentRoom ? (
            <div className={classes.partner}>
              <ChatAvatar partner={currentRoom.partner} size="large" />
              <Typography
                className={classes.title}
                component={"h1"}
                variant="h6"
                align="center"
              >
                {currentRoom.partner.name}
              </Typography>
            </div>
          ) : (
            <Typography
              className={classes.title}
              component={"h1"}
              variant="h3"
              align="center"
            >
              Chat
            </Typography>
          )}

          <Divider />
          {error &&
            Array.isArray(error) &&
            error.map((err, idx) => <p key={idx}>{err}</p>)}
          {currentRoom && <ChatRoom />}
        </Container>
      </div>
    </Container>
  );
}

export default Chats;
