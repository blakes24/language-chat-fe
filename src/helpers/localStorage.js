export function setLocalStorage(key, val) {
  window.localStorage.setItem(key, JSON.stringify(val));
}

export function getFromLocalStorage(key, defaultValue) {
  let value;
  try {
    value = JSON.parse(
      window.localStorage.getItem(key) || JSON.stringify(defaultValue)
    );
  } catch (e) {
    console.error(e);
    value = defaultValue;
  }
  return value;
}
