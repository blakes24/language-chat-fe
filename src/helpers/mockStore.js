import configureStore from "redux-mock-store";

const middlewares = [];
const mockStore = configureStore(middlewares);

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const initialState = {
  users: {
    current: testUser,
    items: [],
    token: "",
    loading: "idle",
    error: null,
  },
  rooms: { items: [], current: null, loading: "idle", error: null },
  messages: { items: [], loading: "idle", error: null },
  partners: { items: [], loading: "idle", error: null },
};

export const store = mockStore(initialState);
