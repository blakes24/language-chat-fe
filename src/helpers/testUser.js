import UserContext from "./UserContext";

const testUser = {
  id: 1,
  name: "Test",
  bio: "stuff about mae",
  email: "test@mail.com",
  imageUrl: "https://cdn.fakercloud.com/avatars/happypeter1983_128.jpg",
  active: true,
  speaks: [
    {
      code: "it",
      language: "Italian",
    },
  ],
  learning: [
    {
      code: "ko",
      language: "Korean",
      level: "2",
    },
  ],
};

const UserProvider = ({
  children,
  user = testUser,
  setToken = () => console.log("set token"),
}) => <UserContext.Provider value={{ user }}>{children}</UserContext.Provider>;

export { UserProvider };
