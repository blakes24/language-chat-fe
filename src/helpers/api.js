import axios from "axios";
import { BASE_URL } from "../config";
import { getFromLocalStorage } from "./localStorage";

/** API Class: methods used to connect to the API. */

let token = getFromLocalStorage("token", "");

class ChatApi {
  static token = token;

  static async request(endpoint, data = {}, method = "get") {
    console.debug("API Call:", endpoint, data, method);
    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${ChatApi.token}` };
    const params = method === "get" ? data : {};
    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      let message = err.response
        ? err.response.data.error.message
        : "Server error: try again later";
      throw Array.isArray(message) ? message : [message];
    }
  }

  /** Create a new user and get auth token. */

  static async register(userData) {
    let res = await this.request(`auth/register`, userData, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get user details. */

  static async getUser(userId) {
    let res = await this.request(`users/${userId}`);
    return res.user;
  }

  /** Verifies that the social token provided is valid */

  static async verify(token) {
    let res = await this.request(`auth/validate`, token, "post");
    return res;
  }

  /** Get gets token to authenticate future requests to api */

  static async getToken(data) {
    let res = data.email
      ? await this.request(`auth/token`, data, "post")
      : await this.request(`auth/social-token`, data, "post");
    this.token = res.token;
    return res.token;
  }

  /** Get all users, filter by speaks or learning language */

  static async getAllUsers(filters) {
    let res = await this.request(`users`, filters);
    return res.users;
  }

  /** Get a users chat rooms, filter by chat partner */

  static async getRooms({ userId, partnerId }) {
    let queryParams = { partner: partnerId };
    let res = await this.request(`users/${userId}/rooms`, queryParams);
    return res;
  }

  /** Creates a chat room for two users data = {user1, user2} */

  static async createRoom(data) {
    let res = await this.request(`rooms`, data, "post");
    return res;
  }

  /** Gets messages, filter by room */

  static async getMessages(roomId) {
    let queryParams = { room: roomId };
    let res = await this.request(`messages`, queryParams);
    return res.messages;
  }

  /** Update user info. */

  static async updateUser(id, data) {
    let res = await this.request(`users/${id}`, data, "patch");
    return res.user;
  }

  /** Update language(speaks). */

  static async updateSpeaks(id, data) {
    let res = await this.request(`languages/speaks/${id}`, data, "patch");
    return res.language;
  }

  /** Update language(learning). */

  static async updateLearning(id, data) {
    let res = await this.request(`languages/learning/${id}`, data, "patch");
    return res.language;
  }

  /** Get a users partners */

  static async getPartners(userId) {
    let res = await this.request(`users/${userId}/partners`);
    return res.partners;
  }

  /** Adds a partner */

  static async addPartner(userId, data) {
    let res = await this.request(`users/${userId}/partners`, data, "post");
    return res;
  }

  /** Deletes a partner */

  static async deletePartner({ userId, partnerId }, data) {
    let res = await this.request(
      `users/${userId}/partners/${partnerId}`,
      data,
      "delete"
    );
    return res;
  }

  /** Deletes a user */

  static async deleteUser(userId, data) {
    let res = await this.request(`users/${userId}`, data, "delete");
    return res;
  }
}

export default ChatApi;
