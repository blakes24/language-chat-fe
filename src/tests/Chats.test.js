import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { store } from "../helpers/mockStore";
import { Provider } from "react-redux";
import Chats from "../components/chat/Chats";

test("renders without crashing", () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <Chats />
      </MemoryRouter>
    </Provider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter>
        <Chats />
      </MemoryRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
