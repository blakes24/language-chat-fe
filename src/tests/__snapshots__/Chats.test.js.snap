// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root makeStyles-root-8 MuiContainer-maxWidthXl"
  >
    <div
      class="makeStyles-chat-10"
    >
      <div
        class="makeStyles-list-9"
      />
      <div
        class="MuiContainer-root makeStyles-chatContainer-12 MuiContainer-maxWidthLg"
      >
        <h1
          class="MuiTypography-root makeStyles-title-11 MuiTypography-h3 MuiTypography-alignCenter"
        >
          Chat
        </h1>
        <hr
          class="MuiDivider-root"
        />
        <p
          class="MuiTypography-root makeStyles-empty-14 MuiTypography-body1 MuiTypography-alignCenter"
        >
          Select a user to start chatting.
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
